functions/logStripeEvent.js

```diff
@@ -0,0 +1,90 @@
+const functions = require('firebase-functions');
+const admin = require('firebase-admin');
+const logDonation = require('./logDonation');
+if (admin.apps.length === 0) {
+  admin.initializeApp();
+}
+
+const DEFAULTCPL = 1.0;
+
+exports.logPaymentIntent = functions.https.onRequest(async (req, res) => {
+  if (!req.body) {
+    res.status(501).send('no data supplied!');
+    return;
+  }
+  let event;
+  try {
+    event = req.body;
+  } catch (err) {
+    console.log(`encountered error while parsing request: ${err}`);
+    res.status(501).send({msg: 'bad data supplied', data: req.body});
+    return;
+  }
+  console.log(`parsing event with id ${event.id}`);
+  switch (event.type) {
+    case 'payment_intent.succeeded':
+      const paymentIntent = event.data.object;
+      if (paymentIntent.description === 'Give Lively / Smart Donations') {
+        console.log(`successful payment for ${paymentIntent.amount}`);
+        this.handlePaymentIntentSucceeded(paymentIntent, event.id);
+      }
+      break;
+    default:
+      // ignore other events
+      break;
+  }
+  res.status(200).send({msg: 'sucessfully received event', data: event});
+  return;
+});
+
+exports.handlePaymentIntentSucceeded = async (intent, id) => {
+  let metadata;
+  let amount = intent.amount/100; // convert from cents to dollars
+  try {
+    metadata = intent.metadata;
+    let coveredByDonor = metadata.transaction_fee_covered_by_donor;
+    if (coveredByDonor) {
+      coveredByDonor = Number(coveredByDonor.replace('$', ''));
+      amount = amount - coveredByDonor;
+    }
+    let splitString;
+    let campaignID;
+    let country;
+    let referralSource;
+    try {
```

Comment:
In general, with Node, it's a best practice to avoid a try/catch for basic validations - mostly for readability and how Node internally handles an error when it's thrown (it stops the program and traces the call stack to find the nearest `catch` block).  

You can write the assignment of the `splitString` variable with `metadata||''.utm_source||''.split('|')` or `metadata?.utm_source||''.split('|')` (using the optional chaining operator).  Both versions will return an array with `['']` so the first assignment of `campaignID` will be `MISSING` since the empty string value is false and all other array values will be `undefined`.  This will avoid the need for the try/catch statement

Reply:
great idea, thank you!